File: compute_con.R
############################################################
############################################################
############## Accoradance and Discordance #################
############################################################
############################################################

compute_con=function(X, u, l){   #create function, X is the design matrix, u and l are the upper and lower thresholds
  N=nrow(X)  #N: regions (in X, rows represent the regions)
  T=ncol(X)  #T:time (in X,columns represent the time)
  
  normalize=function(X){  #create the normalization function
    X = X - mean(X)
    X = X/sd(X)
    return(X)
  }
  
  for(i in 1:N){  #call normalization function for each line of X matrix
    X[i,]=normalize(X[i,])
  }
  ######## X matrix is now normalized ###########
  
  Xu=array(0, dim=dim(X))  # create new Xu,Xl matrices with zeros and dimensions same as X
  Xl=array(0, dim=dim(X))  # same
  
  Xu[X>=u]=1  # apply the thresholds(u) on X and put ones(=1) in Xu for the corresponding positions
  Xl[X<=l]=-1 # apply the thresholds(l) on X and put ones(=1) in Xu for the corresponding positions
  
  Scav=1/T*(Xu%*%t(Xu))
  Scdv=1/T*(Xl%*%t(Xl))
  Sa=Scav+Scdv
  E=diag(Sa)^(-0.5)           # Calculate the energy
  Acc=diag(E)%*%Sa%*%diag(E)  # Calculate the Accordance matrix
  
  Sd=1/T*(Xu%*%t(Xl)+Xl%*%t(Xu))
  Dis=diag(E)%*%Sd%*%diag(E)  # Calculate the Accordance matrix
  
  con=NULL
  
  con$acc=Acc
  con$dis=Dis
  return(con) # return the con (type : con$acc for Acc matrix and con$Dis for Dis matrix)
}

File: compute_con.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%% Accordance and Discordance %%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%to call the function, type con=compute(X,u,l) after defining X,u and l.
%X is the design matrix(time courses, regions), u,l the upper and lower 
%thresholds for Acc and Dis.
%eg. X = (std).*randn(10,20)+(mean); con=compute_con(X,1,1)

function [con] = compute_con(X,u,l)
N=size(X,1);% N: regions (in X, the rows represent the regions)
T=size(X,2);% T: time (in X, the columns represent the time)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%% Normalization function %%%%%%%%%%%%%%%%%%%%%%%%%%
  function normalize(X)
        X = X - mean(X);
        X = X / std(X);
      for i=1:N
        X(i,:) = normalize(X(i,:)); %normalize each line of X matrix 
      end 
  end
%%%%%%%%%%%%%%%%Create zero matrices (Xu,Xl)%%%%%%%%%%%%%%%%%%
Xu=zeros(N,T);
Xl=zeros(N,T);
%%%%%%%%%%%%%%%%%%%%% Apply Thresholds %%%%%%%%%%%%%%%%%%%%%%%
Xu(X>=u)=1; %apply thresholds(u) on X and put ones(=1) in Xu for the corresponding positions 
Xl(X<=l)=-1;%apply theshold l on X and put -1 in Xl for the corresponding positions 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%% ACCORDANCE %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
Scav= 1 / T*(Xu *(Xu'));
Scdv= 1 / T*(Xl *(Xl'));
Sa=Scav + Scdv;
E=diag(Sa).^(-0.5); %energy
Acc=diag(E)* Sa * diag(E); %Calculate the Accordance matrix (Acc)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%% DISCORDANCE %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
Sd = 1 / T*((Xu * (Xl')) + (Xl * (Xu')));
Dis = diag(E) * Sd * diag(E); %Calculate the Discordance matrix (Dis)

con=[];
con.accordance=Acc;
con.discordance=Dis;

end
 